backends/gstreamer/src/player.rs

```diff
@@ -248,8 +214,58 @@ impl PlayerInner {
     }
 }
 
+type PlayerEventObserver = IpcSender<PlayerEvent>;
+struct PlayerEventObserverList {
+    observers: Vec<PlayerEventObserver>,
+}
+
+impl PlayerEventObserverList {
+    fn new() -> Self {
+        Self {
+            observers: Vec::new(),
+        }
+    }
+
+    fn register(&mut self, observer: PlayerEventObserver) {
+        self.observers.push(observer);
+    }
+
+    fn notify(&self, event: PlayerEvent) {
+        for observer in &self.observers {
+            observer.send(event.clone()).unwrap();
+        }
+    }
+}
+
+struct FrameRendererList {
+    renderers: Vec<Arc<Mutex<FrameRenderer>>>,
+}
+
+impl FrameRendererList {
+    fn new() -> Self {
+        Self {
+            renderers: Vec::new(),
+        }
+    }
+
+    fn register(&mut self, renderer: Arc<Mutex<FrameRenderer>>) {
+        self.renderers.push(renderer);
+    }
+
+    fn render(&self, sample: &gst::Sample) -> Result<(), ()> {
+        let frame = frame_from_sample(&sample)?;
+
+        for renderer in &self.renderers {
+            renderer.lock().unwrap().render(frame.clone());
+        }
+        Ok(())
+    }
+}
+
 pub struct GStreamerPlayer {
     inner: RefCell<Option<Arc<Mutex<PlayerInner>>>>,
+    observers: Arc<Mutex<PlayerEventObserverList>>,
+    renderers: Arc<Mutex<FrameRendererList>>,
```

Comment:
do both the lists and the individual renderers need to be behind Arcs and mutexes? I feel like having Vec<FrameRenderer> inside the list may be fine here?

Reply:
Each individual renderer is accessed from the [layout](https://dxr.mozilla.org/servo/source/components/script/dom/htmlmediaelement.rs#1490) and [gstreamer](https://github.com/servo/media/pull/176/files#diff-1d803d24b8ea49dbfeb827b92c7f9c28R261) threads, so I think we need to have them behind Arcs and mutexes.

Reply:
ah, okay then. r=me